Tkt 46: automatically display `equation` context

yanenkoa opened the pull request:
@dbarashev Here's the first version. Still WIP: the formula is updated on every change.

dbarashev commented on demo/summer-2016.html:
```diff
@@ -215,10 +216,10 @@
-  function(ace, langTools, katexPreviewer, highlighter) {
-    var editor = ace.edit("editor");
+require(["ace/ace", "ace/ext/language_tools", "ace/ext/papeeria/context-previewer", "ace/ext/papeeria/highlighter"], function(ace, langTools, katexPreviewer, highlighter) {
+    editor = ace.edit("editor");
```
get back `var` here?

yanenkoa replied:
Is it necessary? It's kind of handy to have this variable in global scope in browser, that lets me experiment right in the browser console.

dbarashev replied:
Well, for development purposes you can do whatever makes sense, but in production code it is better to be safe, even if this particular file is not going to production. One global variable assigmnent here, another read access there which for some reasons slipped through the review, and finally you have code which heavily depends on global variables.

yanenkoa replied:
Yeah, I understand that global variables are bad. So, how about when I will merge this branch, just before that I will put `editor` back in this scope?

dbarashev replied:
add TODO ?

yanenkoa replied:
Yes, sure.

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -7,25 +7,21 @@ define((require, exports, module) ->
-      isVisible: (jqPopoverElement) ->
-        jqPopoverElement.children(".popover").is(":visible")
+      isVisible: (jqPopoverContainer) ->
+        jqPopoverContainer.data().popover.tip().hasClass("in")
```
is it any better than `is(":visible")` ?

yanenkoa replied:
I think it is. The difference is I don't have to specify the jQuery element of the popover itself -- only the element, under which the popover is displayed. So previously, I'd have to write `popoverHandler.isVisible($(".popover-content"))`, whereas with, for example, `show`, I'd have to write `popoverHandler.show($("#formula"))`. So, I think, it's better for both simplicity and consistency.

dbarashev replied:
However, this code looks more fragile and less commonly used. It is fragile because relies on existence of `data().popover.tip()` and some particular class. Should anything break in this chain, or should class name change (e.g. with the new version of Bootstrap), this code is broken. 

With respect to conventional practices, one may expect that visibility of a big element is checked with the test of property "visible", and when a programmer sees this code, he just trusts that it produces the expected result. The same result, produced by checking the class of some particular small element is more difficult to trust -- the reader really gets confused and tries to grasp why this very way was used instead of more conventional one.

yanenkoa replied:
`data().popover.tip()`, as far as I understand, is generated by Bootstrap when creating the popover under this element. So, as long as Bootstrap doesn't change its API, we should be good in this part. As for the `.hasClass("in")`, would it help to change this to `.is(":visible")`? So the whole thing would be `jqPopoverContainer.data().popover.tip().is(":visible")`. 

About possible new version of Bootstrap -- we're already working here with Bootstrap 2, which is obsolete. In Bootstrap 3, the API for a lot of things is slightly different. So a lot of code will break with the new version of Bootstrap, not just this one.

dbarashev replied:
Yes, I know that Bootstrap 3 and 4 change API in incompatible ways and there are many other issues besides popovers. However, the rule still holds in general case: long chain of calls is fragile and error prone. 

Why do you want the tip so much ;) ? This method is supposed to return `true` if popover is visible, the most straightforward way of checking is to check whether the container of all popover's elements is visible. You suggest checking if one small element from the container is visible. Why that element, why not popover-content or popover-title? Will it work in imaginary case when popover has no arrow? You leave so many questions to the reader of this code.

yanenkoa replied:
`jqPopoverContainer.data().popover.tip()` returns the whole jQuery element of popover. It's not just the arrow, as it would seem from the name, it is actually the whole thing. Why they decided to call it that, I have no idea.

dbarashev replied:
damn :) maybe it is yet another argument in favor of more simple code :) ? Long line which does something different from whatever you think it does vs short line which does exactly the same what is says, and both are equivalent

yanenkoa replied:
But they are not. See, the way it was before, which is:
```
isVisible: (jqPopoverElement) ->
  jqPopoverElement.children(".popover").is(":visible")
```

`jqPopoverElement` had to be the actual popover element (**not** `$("#formula")`). So, before calling `isVisible`, I had to get the actual popover element somehow. 

So, I could inspect the page and then pick it manually using jQuery, like `$(".popover-content")`. But in this case, what if there are other popovers? From the popover itself, there is no way to find out its origin. So there is a risk to pick the wrong one.

The only way I found to have the right popover element 100% of the time is the way I used here. So I decided to go with it.

By the way, so far I didn't use this method once in my code, so one of the options is to just remove it altogether :)

dbarashev replied:
Ah, I see. Okay, convinced. However, if this method is unused anyway, just delete it :)

dbarashev commented on lib/ace/ext/papeeria/context-previewer.coffee:
```diff
@@ -0,0 +1,133 @@
+        jqPopoverElement.html(content)
+    }
+
+    initKaTeX = (onLoaded) ->
```
What's the point of duplicating all this stuff? Why don't you just add listener on editor changes and that code which processes the context into existing module?

yanenkoa replied:
Do you mean duplicating the `popoverHandler` or `initKatex?

dbarashev replied:
all the initialization and interface code

yanenkoa replied:
So, I should move `popoverHandler` and `initKatex` to different module and `require` them from both `katex-previewer` and `context-previewer`?

dbarashev replied:
why don't you just stick to katex-previewer module?

yanenkoa replied:
Well katex-previewer works a bit differently. Are you saying I should have just replaced the katex-previewer functionality?

dbarashev replied:
Well, it is the same logical unit. The only difference is when you trigger popover, on explicit action from user or implicitly when cursor gets into the context.

dbarashev replied:
@yanenkoa I don't think that we want to throw away Alt+P completely, we'd rather want both ways to co-exist.

yanenkoa replied:
Not entirely. There's also the difference in the logic of when the popover disappears and where it appears. Also, there is additional logic with updating the popover. I guess I could make some sort of small framework and rewrite all the logic in both katex-previewer and context-previewer in terms of that framework, but wouldn't that be overkill?

dbarashev replied:
You can probably start with the most simple scenario: just trigger popover automatically, leave the remaining part as is? I don't think that you need a framework which willbe used in both modules, I think you need a single module with different ways of triggering and hiding popover.

yanenkoa replied:
Ok, I think I understand you now. Will do.

yanenkoa commented:
@dbarashev pushed another portion of commits. Added popover scrolling with the document when the context is previewed, moved `context-previewer` to `katex-previewer` and did some refactoring.

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+          popoverHandler.destroy(jqFormula())
 
-    destroyPopover = -> popoverHandler.hide($("#formula"))
+        ch.prevContext = currentContext
```
isn't it enough to check `contextPreviewExists` only? If it is `false` and you're in equation, you show popover, if it is `true` and you're not in equation, you hide it

yanenkoa replied:
It is. Will fix.

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+        parseInt(jqEditorContainer.find("div.ace_gutter > div.ace_layer.ace_gutter-layer.ace_folding-enabled > div:nth-child(1)").text())
+
+      getPopoverPosition: (row) ->
+        secondRowSelector = "div.ace_scroller > div > div.ace_layer.ace_text-layer > div:nth-child(2)"
```
You may want to try [`editor.renderer.textToScreenCoordinates`](https://github.com/bardsoftware/ace/blob/internship-2016/lib/ace/virtual_renderer.js#L1434)

yanenkoa replied:
Thx a lot. I should probably look into API more :)

dbarashev replied:
ACE API isn't the best documented thing in the world, unfortunately. On the other side, their code is pretty good and names are logical, so I often just open their code in the editor and search for keywords which I expect to find -- and quite often it works fine :)

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
+
+      getTopmostRowNumber: ->
```
What's this? The same as `editor.getFirstVisibleRow` ?

yanenkoa replied:
Yeah, that's a mistake. Will fix.

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
```
You can use `@` in object's methods.

yanenkoa replied:
It won't work, I tried. The thing is, we pass `ch.handleCurrentContext` as a callback function for Ace event, so it changes what `this` is. That's actually the reason why I named the whole object `ch` instead of `contextHandler`.

dbarashev replied:
I see. Yes, if it runs in the event handler then chance are that`this` is going to be different. However, you can pass a closure as the callback, and call `ch`'s methods from the closure, and in this case everything must be fine with `this`

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
```
So, this hack with context probes and filtering with regexp is okay for the proof of the concept; however in production we should probably use more reliable techniques: either our own or KaTeX's tokenizers. 

Also, it seems that KaTeX can just render unknown macros as text, maybe that's a good trade-off?

dbarashev commented on lib/ace/ext/papeeria/katex-previewer.coffee:
```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    ch = {
```
Leave him his full name pls :)

yanenkoa replied:
Should I still leave the full name, considering I can't use `@` inside this object? It will make some methods really verbose. But I don't insist.
